### ISSUE TITLE
opt: excessive planning time for a multi-region insert statement

### ISSUE THREAD

**Issue by rytaft** [cockroachdb/cockroach#102043]
**Describe the problem**

Planning time is way too long for a multi-region insert statement.

**To Reproduce**

I ran [v23.1.0-beta.2](https://www.cockroachlabs.com/docs/releases/v23.1#v23-1-0-beta-2) on a multi-region roachprod cluster, and produced the following statement bundle: https://drive.google.com/file/d/1NynNawnw3zXQFHsLmkTNKypiLw8l6ziC/view?usp=sharing

It's also possible to see the same behavior with `cockroach demo --nodes 9 --global`:
```
CREATE DATABASE testdb PRIMARY REGION "us-east1" REGIONS "us-west1", "europe-west1";
USE testdb;

DROP TABLE IF EXISTS employees;

CREATE TABLE employees (
	emp_no     INT8 NOT NULL,
	birth_date DATE NOT NULL,
	first_name VARCHAR(14) NOT NULL,
	last_name  VARCHAR(16) NOT NULL,
	gender     CHAR NOT NULL,
	hire_date  DATE NOT NULL,
	email      STRING NOT NULL,
	zip_code   STRING NOT NULL,
	UNIQUE (email),
	UNIQUE (last_name, first_name, zip_code, birth_date),
	PRIMARY KEY (emp_no)
) LOCALITY REGIONAL BY ROW;

CREATE OR REPLACE FUNCTION f_random_text(
    length integer
)
RETURNS text AS
$body$
SELECT string_agg(_char, '')
FROM (SELECT _char FROM (SELECT unnest(string_to_array('A B C D E F G H I J K L M N O P Q R S T U V W X Y Z 0 1 2 3 4 5 6 7 8 9', ' ')) AS _char) chars ORDER BY random() LIMIT $1) charlist;
$body$
LANGUAGE sql;

CREATE OR REPLACE FUNCTION f_random_gender()
RETURNS char AS
$body$
SELECT string_agg(_char, '')
FROM (SELECT _char FROM (SELECT unnest(string_to_array('M F N', ' ')) AS _char) chars ORDER BY random() LIMIT 1) charlist;
$body$
LANGUAGE sql;

CREATE OR REPLACE FUNCTION f_random_date()
RETURNS date AS
$body$
SELECT date(now() - trunc(random() * 365)::int * '1 day'::interval - trunc(random() * 100)::int * '1 year'::interval)
$body$
LANGUAGE sql;

CREATE OR REPLACE FUNCTION f_random_zip_code()
RETURNS text AS
$body$
SELECT string_agg(_char, '')
FROM (SELECT _char FROM (SELECT unnest(string_to_array('0 1 2 3 4 5 6 7 8 9', ' ')) AS _char) chars ORDER BY random() LIMIT 5) charlist;
$body$
LANGUAGE sql;

INSERT INTO employees (
  crdb_region,
  emp_no,
  birth_date,
  first_name,
  last_name,
  gender,
  hire_date,
  email,
  zip_code
)
  SELECT
    CASE trunc(random() * 10) % 3 WHEN 0 THEN 'us-east1' WHEN 1 THEN 'us-west1' ELSE 'europe-west1' END,
    generate_series(0, 100000), 
    (now() - trunc(random() * 365)::int * '1 day'::interval - 20 * '1 year'::interval - trunc(random() * 60)::int * '1 year'::interval)::date, 
    f_random_text(10), 
    f_random_text(10), 
    f_random_gender(), 
    (now() - trunc(random() * 365)::int * '1 day'::interval - trunc(random() * 10)::int * '1 year'::interval)::date,
    f_random_text(10) || '@cockroachlabs.com',
    f_random_zip_code();
```
Now run the following:
```
EXPLAIN ANALYZE INSERT INTO employees
  SELECT
    200001, 
    (now() - trunc(random() * 365)::int * '1 day'::interval - 20 * '1 year'::interval - trunc(random() * 60)::int * '1 year'::interval)::date, 
    f_random_text(10), 
    f_random_text(10), 
    f_random_gender(), 
    (now() - trunc(random() * 365)::int * '1 day'::interval - trunc(random() * 10)::int * '1 year'::interval)::date,
    f_random_text(10) || '@cockroachlabs.com',
    f_random_zip_code();
```
Notice that the planning time is way longer than the execution time:
```
demo@127.0.0.1:26257/testdb> EXPLAIN ANALYZE INSERT INTO employees ...                                                                          info
--------------------------------------
  planning time: 279ms
  execution time: 72ms
...

Time: 355ms total (execution 354ms / network 1ms)
```

**Expected behavior**

There's no reason planning time should be so high. We should figure out where the bottlenecks are and remove them.

Jira issue: CRDB-27236

**Comment by rytaft** [comment]
Note that this seems to be almost entirely due to the UDFs in the query. If I replace the UDFs with hardcoded strings the planning time shrinks significantly:
```
...
  planning time: 3ms
  execution time: 91ms
...
```

**Comment by DrewKimball** [comment]
Maybe due to fetching descriptors for the UDFs?

**Comment by rytaft** [comment]
> Maybe due to fetching descriptors for the UDFs?

Seems likely. Should this go to the schema team?

**Comment by chengxiong-ruan** [comment]
I tried to repro on both `v23.1.0-beta.2` and `v23.1.0-rc.2` with `cockroach demo --nodes 9 --global --insecure` on my laptop, but couldn't repro what @rytaft saw :(
On both versions I got similar results:
By using UDFs I got:
```
  planning time: 6ms
  execution time: 418ms
......
```
By replacing UDFs with hardcoded strings, I got:
```
planning time: 3ms
execution time: 409ms
......
```

steps to repro:
1. start crdb demo with `cockroach demo --nodes 9 --global --insecure`
2. setup schema and functions as described in the ticket
3. run `EXPLAIN ANALYZE` on the same queries as described

Could someone help double check? Just want to gather more data points before digging further. Thanks!
